bugsnag-plugin-android-ndk/src/main/jni/utils/serializer/event_writer.c

```diff
@@ -2,156 +2,726 @@
 
 #include <fcntl.h>
 #include <string.h>
+#include <time.h>
 #include <unistd.h>
 
-#include "../string.h"
+#include "BSG_KSCrashStringConversion.h"
+#include "BSG_KSJSONCodec.h"
 #include "buffered_writer.h"
-#include "utils/seqlock.h"
+#include "event.h"
+#include "internal_metrics.h"
+#include "utils/string.h"
 
-bool bsg_write_breadcrumbs(bugsnag_event *event, bsg_buffered_writer *writer);
-bool bsg_write_feature_flags(bugsnag_event *event, bsg_buffered_writer *writer);
+#define CHECKED(e)                                                             \
+  if ((e) != BSG_KSJSON_OK) {                                                  \
+    goto error;                                                                \
+  }
+
+#define JSON_CONSTANT_ELEMENT(name, value)                                     \
+  bsg_ksjsonaddStringElement(json, name, value,                                \
+                             sizeof(value) - 1 /* remove terminator */)
+#define JSON_LIMITED_STRING_ELEMENT(name, value)                               \
+  bsg_ksjsonaddStringElement(json, (name), (value),                            \
+                             strnlen((value), sizeof((value))))
+
+#define STRING_NOT_EMPTY(s) (*(s) != 0)
+#define STRING_IS_EMPTY(s) (*(s) == 0)
```

Comment:
I feel like these two should be inline functions since there's no inherent type checking possible otherwise.